vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs

```diff
@@ -417,49 +479,110 @@ and
                     projectContextFactory.CreateProjectContext(
-                        FSharpConstants.FSharpLanguageName, projectDisplayName, projectFileName, projectGuid, hierarchy, null, errorReporter)
-
-                let project = projectContext :?> AbstractProject
-
+                              FSharpConstants.FSharpLanguageName, 
+                              projectDisplayName, 
+                              projectFileName, 
+                              projectGuid, 
+                              hierarchy, 
+                              null, 
+                              errorReporter) :?> AbstractProject
+                | _ as p -> p
+
+            match project :> obj with
+            | :? IWorkspaceProjectContext as projectContext ->
                 this.SyncProject(project, projectContext, site, workspace, forceUpdate=false, userOpName=userOpName)
-                site.AdviseProjectSiteChanges(FSharpConstants.FSharpLanguageServiceCallbackName, 
-                                              AdviseProjectSiteChanges(fun () -> this.SyncProject(project, projectContext, site, workspace, forceUpdate=true, userOpName="AdviseProjectSiteChanges."+userOpName)))
-                site.AdviseProjectSiteClosed(FSharpConstants.FSharpLanguageServiceCallbackName, 
-                                             AdviseProjectSiteChanges(fun () -> 
+                site.AdviseProjectSiteChanges(FSharpConstants.FSharpLanguageServiceCallbackName,
+                                                AdviseProjectSiteChanges(fun () -> this.SyncProject(project, projectContext, site, workspace, forceUpdate=true, userOpName="AdviseProjectSiteChanges."+userOpName)))
+                site.AdviseProjectSiteClosed(FSharpConstants.FSharpLanguageServiceCallbackName,
+                                                AdviseProjectSiteChanges(fun () ->
                                                 projectInfoManager.ClearInfoForProject(project.Id)
                                                 optionsAssociation.Remove(projectContext) |> ignore
                                                 project.Disconnect()))
-                for referencedSite in ProjectSitesAndFiles.GetReferencedProjectSites (site, this.SystemServiceProvider) do
-                    let referencedProjectId = setup referencedSite                    
-                    project.AddProjectReference(ProjectReference referencedProjectId)
+            | _ -> ()
+            for referencedSite in ProjectSitesAndFiles.GetReferencedProjectSites (site, this.SystemServiceProvider) do
+                let referencedProjectId = setup referencedSite
+                project.AddProjectReference(ProjectReference referencedProjectId)
 
             projectId
         setup (siteProvider.GetProjectSite()) |> ignore
 
+    member private __.ProvideProjectSiteProvider(workspace:Workspace, project:Project) =
```

Comment:
I think this should be a helper function either outside this class (to prove that it is logically independent of any of the constructs accessible in the class), or as a ``let`` function in the class ``let`` section, to help show what data it depends in the initialization sequence.  Perhaps it can even go in ProjectSitesAndFiles.fs where two other ways of creating IProjectSite are implemented (for scripts and standalone-files).